atc/db/team.go

```diff
@@ -163,6 +164,25 @@ func (t *team) FindWorkerForVolume(handle string) (Worker, bool, error) {
 	}))
 }
 
+func (t *team) FindWorkersForResourceCache(rcId int) ([]Worker, error) {
+	return getWorkers(
+		t.conn, workersQuery.
+			Join("volumes v ON v.worker_name = w.name").
+			Join("worker_resource_caches wrc ON v.worker_resource_cache_id = wrc.id").
+			Join("resource_caches rc ON rc.id = wrc.resource_cache_id").
+			Where(sq.And{
+				sq.Eq{"rc.id": rcId},
+				sq.Eq{"w.state": "running"},
+			}).
+			Where(sq.Or{
+				sq.Eq{
```

Comment:
I think it might be a little more complicated than this. So I'm pretty sure that currently, if a team has a team worker then we will only ever use the team worker. This part of the code is where we will return only the team workers if there are any compatible team workers

https://github.com/concourse/concourse/blob/81c39e21d445069ebb95d8d1f63c9ed63eaf1a36/atc/worker/pool.go#L88-L92

With that in mind, I think we need to check to see if this team has any team workers and if it does, then we pull off the workers that have a resource cache that matches. If the team does not have any team workers, then we can grab a resource cache off any general workers (workers that are not part of a team and not tagged).

Reply:
Now I moved the team and tags matching to `atc/worker/pool.go`, so that it reuse same logic of `pool.allSatisfying()` for matching team and tags.